DataParallelTable.lua

```diff
@@ -494,106 +371,309 @@ function DataParallelTable:name()
+         else
+            size = tensors[i]:size()
+         end
+      end
+   end
+   return size
 end
 
-function DataParallelTable:_calculateSliceRange(tensor, id, total)
-   local outerDim = tensor:size(self.dimension)
-   local eltsPerMod = torch.round( outerDim / #self.modules )
-   local rangeStart = (id - 1) * eltsPerMod + 1
-   local rangeEnd = rangeStart + eltsPerMod - 1
-   if id == total then
-      rangeEnd = outerDim
+function DataParallelTable:_distribute(src, dst)
+   for i = 1, #self.gpuAssignments do
+      cutorch.setDevice(self.gpuAssignments[i])
+      dst[i] = self:_distributeTensorRecursive(src, dst[i], i, #self.gpuAssignments)
    end
-   self.batchSize = outerDim -- TODO: this is a hack to propagate batchSize to line 494
-                             --       but might not be generic enough
-   return {rangeStart, rangeEnd}
 end
 
 -- _distributeTensorRecursive - if the src is a tensor then the function slices
 -- it long self.dimension and copies each portion into each child module.
 -- Otherwise it does a recursive call on tables.
-function DataParallelTable:_distributeTensorRecursive(src, dst,
-   srcGpuid, srcIndex, dstGpuid, dstIndex, nModules)
-   if (torch.type(src) == 'table') then
+function DataParallelTable:_distributeTensorRecursive(src, dst, idx, n)
+   if torch.type(src) == 'table' then
       if torch.type(dst) ~= 'table' or #src ~= #dst then
          dst = {}
       end
 
       -- Recurse on the table
       for i = 1, #src do
-         dst[i] = self:_distributeTensorRecursive(src[i], dst[i], srcGpuid,
-         srcIndex, dstGpuid, dstIndex, nModules)
-      end
-
-   elseif torch.type(src):find('torch%..+Tensor') then
-      if (dst == nil or torch.type(dst) ~= 'torch.CudaTensor') then
-         -- Allocate only on startup or when input table structure changes.
-         -- Otherwise we will just resize the tensor below.
-         setDevice(dstGpuid)
-         dst = torch.CudaTensor()
+         dst[i] = self:_distributeTensorRecursive(src[i], dst[i], idx, n)
       end
+      return dst
+   end
 
-      -- Split the tensor
-      assert(torch.typename(src) == 'torch.CudaTensor')
-      local slice = src[{self:_calculateSliceRange(src, dstIndex, nModules)}]
+   assert(torch.isTensor(src), 'input must be a tensor or table of tensors')
+   assert(src:type() == 'torch.CudaTensor' or src:type() == 'torch.FloatTensor',
+      'input must be a CUDA or Float tensor')
 
-      if not dst:isSameSizeAs(slice) then
-         setDevice(dstGpuid)
-         dst:resizeAs(slice)
-      end
+   dst = torch.type(dst) == 'torch.CudaTensor' and dst or torch.CudaTensor()
 
-      asyncGPUCopy(dst, slice)  -- dst, src
+   local index, size = sliceRange(src:size(self.dimension), idx, n)
+   if size == 0 then
+      dst:resize(0)
    else
-      error('input must be a nested table of tensors!')
+      local slice = src:narrow(self.dimension, index, size)
+      dst:resize(slice:size()):copyAsync(slice)
```

Comment:
If copies are made from the GPU, and the stream is non-default, additional sync is required. Torch runs D2D copies on the source device, and destination devices will start calculating before copies are finished. You can profile attached file to observe this behaviour. 
[copytest.txt](https://github.com/torch/cunn/files/96563/copytest.txt)
![copytest](https://cloud.githubusercontent.com/assets/15841449/12432037/b4d81a3c-bead-11e5-9480-1d49569e7a8c.png)